File: experiments_ros2/src/topics/examples/pub_msg_custom.cpp
#include <chrono>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include "custom_interfaces/msg/msg_nuevo.hpp"

class PublisherNode : public rclcpp::Node
{
public:
  PublisherNode()
    : Node("custom_publisher")
  {
    publisher_ = this->create_publisher<custom_interfaces::msg::MsgNuevo>("custom_topic", 10);

    timer_ = this->create_wall_timer(
      std::chrono::milliseconds(1000),
      std::bind(&PublisherNode::publishMessage, this));
  }

private:
  void publishMessage()
  {
    auto message = custom_interfaces::msg::MsgNuevo();
    message.num = number_++;

    RCLCPP_INFO(this->get_logger(), "Publicando: campo num = '%d'",(int)message.num);

    publisher_->publish(message);
  }

  rclcpp::Publisher<custom_interfaces::msg::MsgNuevo>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  int number_ = 0;
};

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PublisherNode>());
  rclcpp::shutdown();
  return 0;
}

File: exp_ros2_py/exp_ros2_py/pub_msg_custom.py
import rclpy
from rclpy.node import Node

from custom_interfaces.msg import MsgNuevo                     


class MinimalPublisher(Node):

    def __init__(self):
        super().__init__('publicador_msg_personalizado')
        self.publisher_ = self.create_publisher(MsgNuevo, 'topic_msg_pers', 10) 
        timer_period = 0.5
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.i = 0

    def timer_callback(self):
        msg = MsgNuevo()                                                
        msg.num = self.i                                           
        self.publisher_.publish(msg)
        self.get_logger().info('Publicando: "%d"' % msg.num)       
        self.i += 1


def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    rclpy.spin(minimal_publisher)

    minimal_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()